### PR TITLE
In-memory implementation of etcd client for testing

### LINKED ISSUE #40: etcd mock client depends on etcd server packages
kv/etcd/mock.go depends on etcd server packages. The server package in turn pulls in `go.opentelemetry.io/otel` packages. This complicates using dskit in Tempo that pulls in conflicting versions of these otel packages.

Example from trying to use newer versions of dskit in Tempo:

```
$ go mod vendor
go: downloading github.com/google/flatbuffers v2.0.0+incompatible
go: downloading github.com/go-test/deep v1.0.7
go: finding module for package go.opentelemetry.io/otel/semconv
github.com/grafana/tempo/pkg/ring imports
        github.com/grafana/dskit/kv imports
        github.com/grafana/dskit/kv/etcd imports
        go.etcd.io/etcd/server/v3/embed imports
        go.opentelemetry.io/otel/semconv: package go.opentelemetry.io/otel/semconv provided by go.opentelemetry.io/otel at latest version v0.20.0 but not at required version v1.0.0-RC2
```

A library attempting to be lightweight like dskit shouldn't be pulling in otel packages. We should refactor the mock etcd client (which we use for unit tests) in order to avoid pulling in the etcd server.

### PR DESCRIPTION
**Pull request by 56quarters**
Signed-off-by: Nick Pillitteri <nick.pillitteri@grafana.com>

**What this PR does**:

Create an in-memory version of an Etcd v3 client for use testing
the Etcd implementation of kv.Client. Because this client is only
for unit testing another part of dskit, it does not implement all
operations that the real client does - only enough to pass our tests.

The benefit of this in-memory client is that dskit can drop its
dependency on the entire Etcd server which was used in-process for
testing previously. This meant that consumers of dskit inherited
a dependency on the Etcd server and all its dependencies which
caused version conflicts in e.g. Tempo.

**Which issue(s) this PR fixes**:

Fixes #40

**Checklist**
- [X] Tests updated
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: kv/etcd/mock.go [hunk 1]
```diff
@@ -1,71 +1,480 @@
...
+	close(ch)
+}
+
+// sendEvent writes an event to all currently registered consumers. The consumer
+// channels are each read by a goroutine that filters the event and sends it to
+// the caller of the Watch method.
+func (m *mockKV) sendEvent(e clientv3.Event) {
+	m.consumersMtx.Lock()
+	for ch := range m.consumers {
+		ch <- e
```
Comment (replay) [inline kv/etcd/mock.go:144]: should this be a non-blocking write? the channels created in `createConsumer` have a buffer of `0`, so a bad consumer could block this.

Comment (56quarters) [reply kv/etcd/mock.go:144]: Good idea. I think it's probably a good idea to use channels with a non-zero buffer as well. WDYT?

Comment (56quarters) [reply kv/etcd/mock.go]: Fixed in c8424f3ad074264a691b08ec6e1bb3aa1a9e54f2

Comment (replay) [reply kv/etcd/mock.go]: > I think it's probably a good idea to use channels with a non-zero buffer as well. WDYT?

yes, definitely

File: kv/etcd/mock.go [hunk 2]
```diff
@@ -1,71 +1,498 @@
...
+		eventsMtx: sync.Mutex{},
+	}
+
+	return kv
+}
+
+// mockKV is an in-memory implementation of an Etcd client.
+//
+// This implementation has many limitations compared to the real client since it
+// only exist to be used by the Etcd kv.Client implementation during unit tests. As
```
Comment (trevorwhitney) [inline kv/etcd/mock.go:58]: typo
```suggestion
// only exists to be used by the Etcd kv.Client implementation during unit tests. As
```

File: kv/etcd/mock_test.go [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+		kv := newMockKV()
+		kv.values = map[string]mvccpb.KeyValue{
+			string(pair1.Key): pair1,
+			string(pair2.Key): pair2,
+			string(pair3.Key): pair3,
+		}
+		res, err := kv.Get(context.Background(), "/f", clientv3.WithPrefix())
+
+		require.NoError(t, err)
+		assert.ElementsMatch(t, []*mvccpb.KeyValue{&pair1, &pair3}, res.Kvs)
```
Comment (trevorwhitney) [inline kv/etcd/mock_test.go:70]: small nit, renaming these variables makes the assertion a little more immediately clear.

```suggestion
		fooPair := mvccpb.KeyValue{
			Key:   []byte("/foo"),
			Value: []byte("1"),
		}
		bazPair := mvccpb.KeyValue{
			Key:   []byte("/baz"),
			Value: []byte("2"),
		}
		firstPair := mvccpb.KeyValue{
			Key:   []byte("/first"),
			Value: []byte("3"),
		}

		kv := newMockKV()
		kv.values = map[string]mvccpb.KeyValue{
			string(pair1.Key): fooPair,
			string(pair2.Key): bazPair,
			string(pair3.Key): firstPair,
		}
		res, err := kv.Get(context.Background(), "/f", clientv3.WithPrefix())

		require.NoError(t, err)
		assert.ElementsMatch(t, []*mvccpb.KeyValue{&fooPair, &firstPair}, res.Kvs)
```

File: kv/etcd/mock_test.go [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+	})
+
+	t.Run("empty prefix", func(t *testing.T) {
+		pair1 := mvccpb.KeyValue{
+			Key:   []byte("/foo"),
+			Value: []byte("1"),
+		}
```
Comment (trevorwhitney) [inline kv/etcd/mock_test.go:77]: nit: I would add another pair to the test case as an empty prefix (based on my understanding of the code) should match all.

File: kv/etcd/mock_test.go [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+		pair2 := mvccpb.KeyValue{
+			Key:   []byte("/baz"),
+			Value: []byte("2"),
+		}
+		pair3 := mvccpb.KeyValue{
+			Key:   []byte("/first"),
+			Value: []byte("3"),
+		}
```
Comment (trevorwhitney) [inline kv/etcd/mock_test.go:143]: same as above with the rename suggestions to `fooPair`, `bazPair`, and `firstPair`

File: kv/etcd/mock_test.go [hunk 6]
```diff
@@ -0,0 +1,394 @@
...
+func TestMockKV_Watch(t *testing.T) {
+	// setupWatchTest spawns a goroutine to watch for events matching a particular key
+	// emitted by a mockKV. Any observed events are sent to callers via the returned channel.
+	// The goroutine can be stopped using the return cancel function and waited for using the
+	// returned wait group.
+	setupWatchTest := func(key string, prefix bool) (*mockKV, context.CancelFunc, chan *clientv3.Event, *sync.WaitGroup) {
+		kv := newMockKV()
+		// Use a condition to make sure the goroutine has started using the watch before
+		// we do anything to the mockKV that would emit an event the watcher is expecting
+		cond := sync.NewCond(&sync.Mutex{})
```
Comment (trevorwhitney) [inline kv/etcd/mock_test.go:311]: did not know about these, cool!

**Review by trevorwhitney — APPROVED** [review]
A few small nits, but LGTM! Looks great!

File: kv/etcd/mock.go [hunk 7]
```diff
@@ -1,71 +1,498 @@
...
+	succeeded := m.evalCmps(cmps)
+
+	var toRun []clientv3.Op
+	if succeeded {
+		toRun = thens
+	} else {
+		toRun = elses
+	}
+
+	responses := make([]*etcdserverpb.ResponseOp, len(toRun))
```
Comment (replay) [inline kv/etcd/mock.go:328]:
```suggestion
	responses := make([]*etcdserverpb.ResponseOp, 0, len(toRun))
```

Otherwise this will allocate the slice with the given length, not only the given capacity. Further down we append to this slice, the appended values would then end up behind `len(toRun)` empty values.

File: kv/etcd/mock.go [hunk 8]
```diff
@@ -1,71 +1,498 @@
...
+	for _, o := range toRun {
+		_, err := m.doInternal(o)
+		responses = append(responses, &etcdserverpb.ResponseOp{Response: nil})
```
Comment (replay) [inline kv/etcd/mock.go:331]: should this append the response returned by `m.doInternal()`?

File: kv/etcd/mock.go [hunk 9]
```diff
@@ -1,71 +1,498 @@
...
+		if err != nil {
```
Comment (replay) [inline kv/etcd/mock.go:332]: personally i'd put the error check directly after the call-site which potentially returned an error... but in this specific case I think it makes no difference.

File: kv/etcd/mock.go (hunk 8 above)
Comment (56quarters) [reply kv/etcd/mock.go:331]: I think they're different types and `Response` was difficult to work with but I'll double check.

Comment (56quarters) [reply kv/etcd/mock.go:331]: Yeah, the `ResponseOp` requires response types to implement a private interface and and the types used go through a few `type` aliases. I don't think it's possible to actually construct something that'll work here and it's not required by the calling code.

Comment (replay) [reply kv/etcd/mock.go]: ok, nvm

**Review by replay — APPROVED** [review]
LGTM
